scripts/backfill_errors.py

```diff
@@ -0,0 +1,427 @@
+        event.http_method,
+        event.http_referer,
+        event.tags_key,
+        event.tags_value,
+        event.contexts_key,
+        event.contexts_value,
+        event.transaction_name or "",
+        span_id,
+        trace_id,
+        event.partition or 0,
+        event.offset or 0,
+        event.message_timestamp,
+        event.retention_days,
+        event.deleted,
+        event.group_id,
+        UUID(event.primary_hash),
+        event.received or 0,
+        event.message or "",
+        event.title or "",
+        event.culprit or "",
+        event.level,
+        event.location,
+        event.version,
+        event.type or "",
+        event.exception_stacks_type,
+        event.exception_stacks_value,
+        event.exception_stacks_mechanism_type,
+        event.exception_stacks_mechanism_handled,
+        event.exception_frames_abs_path,
+        event.exception_frames_colno,
+        event.exception_frames_filename,
+        event.exception_frames_function,
+        event.exception_frames_lineno,
+        event.exception_frames_in_app,
+        event.exception_frames_package,
+        event.exception_frames_module,
+        event.exception_frames_stack_level,
+        event.sdk_integrations,
+        event.modules_name,
+        event.modules_version,
+    )
+
+    return {
+        name: value
+        for name, value in zip(Error.field_names(), error._asdict().values())
+    }
+
+
+def backfill_errors() -> None:
+    events_storage = get_writable_storage(StorageKey.EVENTS)
+    errors_storage = get_writable_storage(StorageKey.ERRORS)
+
+    orderby_desc = "project_id DESC, timestamp DESC, event_id DESC"
+    orderby_asc = "project_id ASC, timestamp ASC, event_id ASC"
+
+    cluster = events_storage.get_cluster()
+
+    clickhouse = cluster.get_query_connection(ClickhouseClientSettings.MIGRATE)
+
+    events_table_name = events_storage.get_table_writer().get_schema().get_table_name()
+    errors_table_name = errors_storage.get_table_writer().get_schema().get_table_name()
+
+    try:
+        last_event_id, last_event_project, last_event_timestamp = clickhouse.execute(
+            f"""
+                SELECT replaceAll(toString(event_id), '-', ''), project_id, timestamp FROM {errors_table_name}
+                WHERE type != 'transaction'
+                AND NOT deleted
+                ORDER BY {orderby_asc}
+                LIMIT 1
+            """
+        )[0]
+        print(
+            f"Error data was found; starting migration from {last_event_id} {last_event_project} {last_event_timestamp}"
+        )
+    except IndexError:
+        last_event_id, last_event_project, last_event_timestamp = None, None, None
+
+    src_columns = ", ".join(
+        [cast(str, escape_identifier(field)) for field in Event.field_names()]
+    )
+
+    events_migrated = 0
+
+    while True:
+        event_conditions = ""
+        if last_event_id and last_event_project and last_event_timestamp:
```

Comment:
I don't think this condition will do what you want if you are not also iterating on all projects since while the table is sorted by project, timestamp and event it, project are filled in in parallel.

Example:
This is the content of the events table:
```
project.   timestamp.    event_id
P1             1                     A
P1.            2                    B
P1.            3                     C
P2.            1                     X
P2             2                     Y
P3             1                     Z
```

And  this is what the errors table contains
```
project.   timestamp.    event_id
P1.            2                    B
P1.            3                     C
P2             2                     Y
```

Now when you look at last event id and last project id and last event timestamp from the errors table you will get
```
P1.            2                    B
```

Now this condition will start scanning the events table from `P1,2,B` down. And copy `P1,1,A` into errors. 
`P2,1,X` and `P3, 1 Z` are not going to be copied.

Reply:
Thanks I missed this case earlier. I switched the order of timestamp/project in the orderby clause to address this. Timestamp needs to be first to ensure we don't miss any events with larger project IDs that come in later. I don't know what impact it will have on performance, but it seems the only correct option here to ensure we do not miss events.